### DISCUSSION TITLE
max volume

### DISCUSSION THREAD

**Discussion by kopmand** [pschatzmann/ESP32-A2DP#527] [Q&A] [answered]
Hi.
First of all, thank you for this awsome library.

I could not find any answers in the wiki or in the discussions, so i am trying here.

Qestion:
Are there any possibility to set a maximum volume?
I am using af a PCM5102 DAC decoder module and when i turn the volume to max the sound gets very distorted.
By default get_volume() returns values between 0 and 127 (0-100%)

So i was thinking, would it be possible to change this behavior to for example 0 -> 127 = 0 -> 80% ?

Regards Stefan.

**Comment by harryberlin** [comment]
I'm using such Hifi DAC device on Raspberry.
https://github.com/harryberlin/AVR-IBus.public

Did add 150 Ohm Resistors between Audio R+ & GND and L+ & GND to solve the audio overdrive issue.

**Reply by kopmand** [reply]
That is a great solution.
But in my project i am using 2 libraries, ESP32-A2DP and ESP32-audioI2S.
ESP32-audioI2S does not distort the sound at maximum volume, so i need a solution where i can use both libraries at max volume.
I think that if i ad the resistors to the output, the sound would be to low when i use ESP32-audioI2S ?

**Comment by XPMonline** [comment]
Hi Stefan,
 simply use arduino map() function. For example
`  volume map(volume,0,127,0,80);`

**Reply by kopmand** [reply]
I am actually already using the map() function, and it is working perfect on the "device"
But when i am controlling the volume from my phone, get_volume() overrules the map() function.

**Comment by harryberlin** [comment]
Did read in other Forum, Bridge between sck (if not used) and gnd can solve sound disruption.

**Reply by kopmand** [reply]
I already did that :-)
Without the bridge i doesen't even play.

**Comment by XPMonline** [comment]
As Mr. Schatzmann wrote in another topic: 

> As a side note I just wanted to highlight that you could easily implement a "volume control" by applying a multiplier on the sound data in the stream reader callback that you set via set_stream_reader() e.g.
> 
> multiplying by 0 is muting
> multiplying by 0.5 is giving a lower volume

// In the setup function:
`a2dp_sink.set_stream_reader(read_data_stream);`

// Then somewhere in your sketch:
```
void read_data_stream(const uint8_t *data, uint32_t length)
{
      int16_t *ptr16 = (int16_t*) data; 
      uint21_t len16 = length/2;
      for(int j=0;j<len16;j++){
           ptr16[j] = 0.5 * ptr16[j];
      }
}
```

**Reply by kopmand** [reply]
That looks exactly like the thing i am missing :-)
I tried it, but i get this compile error:     'uint21_t' was not declared in this scope

I am not skilled at programming at all.
Can you tell me what i am doing wrong?
Here are my sketch, based on "bt_music_receiver_simple" with the void read_data_stream()


#include "BluetoothA2DPSink.h"

BluetoothA2DPSink a2dp_sink;

void setup() {
    i2s_pin_config_t my_pin_config = {
      .mck_io_num = I2S_PIN_NO_CHANGE,
      .bck_io_num = 23,
      .ws_io_num = 21,
      .data_out_num = 22,
      .data_in_num = I2S_PIN_NO_CHANGE
    };
    a2dp_sink.set_pin_config(my_pin_config);
    a2dp_sink.set_auto_reconnect(true);
    a2dp_sink.set_stream_reader(read_data_stream);
    a2dp_sink.start("MyMusic"); 
}

void loop() {
  delay(1000); // do nothing
}

void read_data_stream(const uint8_t *data, uint32_t length)
{
      int16_t *ptr16 = (int16_t*) data; 
      uint21_t len16 = length/2;
      for(int j=0;j<len16;j++){
           ptr16[j] = 0.5 * ptr16[j];
      }
}

**Reply by XPMonline** [reply]
Try replace it with uint32_t.  In my project I also use a solution with resistor placement at the input of the power amplifier.
I have not tested this solution with stream reader modification

**Reply by kopmand** [reply]
uint32_t did not help...

**Comment by pschatzmann** [comment, marked as answer]
Why don't you just implement your own A2DPVolumeControl? 
There you can easily limit the value.

Or with the latest release you can wrap the I2SStream in a VolumeStream using the AudioTools

**Reply by vivekprojects1** [reply]
I tried this but the problem still remains. There is no distortion up to about 85% of volume on the phone but as over 85% the sound distorts badly.

**Reply by vivekprojects1** [reply]
This is when using an ESP32 and the external DAC is a TDA1387 X8.

Will appreciate if you can suggest the complete sketch.

Thank you.

**Reply by pschatzmann** [reply]
Just have a look at the DefaultVolumeControl class in [A2DPVolumeControl.h](https://github.com/pschatzmann/ESP32-A2DP/blob/main/src/A2DPVolumeControl.h): you would just limit the volumeFactor to a lower limit...

**Reply by pschatzmann** [reply]
I have added [a constructor which accespts a max value parameter](https://pschatzmann.github.io/ESP32-A2DP/html/class_a2_d_p_default_volume_control.html). 
This should allow you to limit the max volume...

**Reply by vivekprojects1** [reply]
Hi tried the methods above but the out put still distorts when the volume level on the phone is raised.
Then tried  

audio.volume(0.1); 

Any value above 0.1 results in distortion.

With this the line in code the sound no longer distorts but the quality is poor.

Are there any other changes that could be made in the program as have tried adding capacitors on the power line for filtering and even feeding the DAC from a different power supply just in case the distortion is being introduced on the power lines but to no effect.

**Reply by pschatzmann** [reply]
I rather think this is a hardware problem. Test first with one of the [tested dacs](https://github.com/pschatzmann/arduino-audio-tools/wiki/External-DAC)!
If this is working, follow the [instructions for your TDA1387](https://github.com/moode-player/protodac/blob/main/docs/protodac_tda1387_x8.md): Your dac should really support the full 16 bit data range properly to be useful...
